source/common/ssl/context_impl.cc

```diff
@@ -137,35 +147,53 @@ bssl::UniquePtr<SSL> ContextImpl::newSsl() const {
   return bssl::UniquePtr<SSL>(SSL_new(ctx_.get()));
 }
 
-bool ContextImpl::verifyPeer(SSL* ssl) const {
-  bool verified = true;
+int ContextImpl::verifyCallback(int ok, X509_STORE_CTX* store) {
```

Comment:
`error_depth` is checked so that we verify hash and/or SAN only in the leaf certificate and bail out early for intermediates, etc. I'll add comment for that.

Indeed, I could use `SSL_CTX_set_cert_verify_callback`, but it seems a bit more "scary" than `verify_callback` and it would result in virtually the same code in the callback, anyway. The only advantage of `SSL_CTX_set_cert_verify_callback` seems to be `callback_arg`, which could be used instead of `SSL_CTX_set_app_data`.

Would you prefer if I used it here? It seems that both methods would be eliminated once you move away from X509, right? I definitely couldn't use `X509_verify_cert` then, so `verify_callback` might be actually more future proof?